<a href="https://colab.research.google.com/github/lemonbuilder/EXPLORATION/blob/main/E-01/%5BE_01%5D.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# [E-01] 분류
###### (세 문제 풀기)



---
##01. load_digits : 손글씨 이미지 판별

(주절주절 써주어야겠구먼... 이걸 하는 이유, 쟁점사항, 궁금한 점??? + 데이터 셋팅 부연)

In [60]:
''' 데이터 셋팅 '''
# 필요한 모듈 import
from sklearn.datasets import load_digits
from sklearn.model_selection import train_test_split
from sklearn.metrics import classification_report
import pandas as pd

# 데이터 준비
hw = load_digits()

# 데이터 이해
hw_data = hw.data
print(hw_data.shape) 

hw_label = hw.target
print(hw_label.shape)

hw.target_names    # array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9])
hw.keys()     # ['data', 'target', 'frame', 'feature_names', 'target_names', 'images', 'DESCR']


(1797, 64)
(1797,)


dict_keys(['data', 'target', 'frame', 'feature_names', 'target_names', 'images', 'DESCR'])

In [61]:
hw_df = pd.DataFrame(data=hw_data, columns=hw.feature_names)
hw_df["label"] = hw.target
hw_df

pixel_0_0  pixel_0_1  pixel_0_2  pixel_0_3  pixel_0_4  pixel_0_5  \
0           0.0        0.0        5.0       13.0        9.0        1.0   
1           0.0        0.0        0.0       12.0       13.0        5.0   
2           0.0        0.0        0.0        4.0       15.0       12.0   
3           0.0        0.0        7.0       15.0       13.0        1.0   
4           0.0        0.0        0.0        1.0       11.0        0.0   
...         ...        ...        ...        ...        ...        ...   
1792        0.0        0.0        4.0       10.0       13.0        6.0   
1793        0.0        0.0        6.0       16.0       13.0       11.0   
1794        0.0        0.0        1.0       11.0       15.0        1.0   
1795        0.0        0.0        2.0       10.0        7.0        0.0   
1796        0.0        0.0       10.0       14.0        8.0        1.0   

      pixel_0_6  pixel_0_7  pixel_1_0  pixel_1_1  ...  pixel_6_7  pixel_7_0  \
0           0.0        0.0        0.0        0.0  ...        0.0        0.0   
1           0.0        0.0        0.0        0.0  ...        0.0        0.0   
2           0.0        0.0        0.0        0.0  ...        0.0        0.0   
3           0.0        0.0        0.0        8.0  ...        0.0        0.0   
4           0.0        0.0        0.0        0.0  ...        0.0        0.0   
...         ...        ...        ...        ...  ...        ...        ...   
1792        0.0        0.0        0.0        1.0  ...        0.0        0.0   
1793        1.0        0.0        0.0        0.0  ...        0.0        0.0   
1794        0.0        0.0        0.0        0.0  ...        0.0        0.0   
1795        0.0        0.0        0.0        0.0  ...        0.0        0.0   
1796        0.0        0.0        0.0        2.0  ...        0.0        0.0   

      pixel_7_1  pixel_7_2  pixel_7_3  pixel_7_4  pixel_7_5  pixel_7_6  \
0           0.0        6.0       13.0       10.0        0.0        0.0   
1           0.0        0.0       11.0       16.0       10.0        0.0   
2           0.0        0.0        3.0       11.0       16.0        9.0   
3           0.0        7.0       13.0       13.0        9.0        0.0   
4           0.0        0.0        2.0       16.0        4.0        0.0   
...         ...        ...        ...        ...        ...        ...   
1792        0.0        2.0       14.0       15.0        9.0        0.0   
1793        0.0        6.0       16.0       14.0        6.0        0.0   
1794        0.0        2.0        9.0       13.0        6.0        0.0   
1795        0.0        5.0       12.0       16.0       12.0        0.0   
1796        1.0        8.0       12.0       14.0       12.0        1.0   

      pixel_7_7  label  
0           0.0      0  
1           0.0      1  
2           0.0      2  
3           0.0      3  
4           0.0      4  
...         ...    ...  
1792        0.0      9  
1793        0.0      0  
1794        0.0      8  
1795        0.0      9  
1796        0.0      8  

[1797 rows x 65 columns]

In [62]:
# 데이터 구분 (X=input, y=target)
X_train, X_test, y_train, y_test = train_test_split(hw_data, hw_label, test_size=0.2, random_state=42)

print('X_train 개수: ', len(X_train),', X_test 개수: ', len(X_test))
print(X_train.shape, y_train.shape)

X_train 개수:  1437 , X_test 개수:  360
(1437, 64) (1437,)


**[학습01.] Decision Tree**

In [90]:
''' 학습 셋팅 '''
from sklearn.tree import DecisionTreeClassifier
decision_tree = DecisionTreeClassifier(random_state=32)

# 학습
decision_tree.fit(X_train, y_train)

# 예측
y_pred_01 = decision_tree.predict(X_test)

# 정확도
from sklearn.metrics import accuracy_score
accuracy = accuracy_score(y_test, y_pred_01)
print(f'Decision Tree의 정확도는 {accuracy}') 

Decision Tree의 정확도는 0.8527777777777777


**[학습02.] Random Forest**

In [91]:
''' 학습 셋팅 '''
from sklearn.ensemble import RandomForestClassifier
random_forest = RandomForestClassifier(random_state=32)

# 학습
random_forest.fit(X_train, y_train)

# 예측
y_pred_02 = random_forest.predict(X_test)

# 정확도
accuracy = accuracy_score(y_test, y_pred_02)
print(f'Random Forest의 정확도는 {accuracy}') 

Random Forest의 정확도는 0.9805555555555555


**[학습03.] Support Vector Machine(SVM)**

In [92]:
''' 학습 셋팅 '''
from sklearn import svm
svm_model = svm.SVC()

# 학습
svm_model.fit(X_train, y_train)

# 예측
y_pred_03 = svm_model.predict(X_test)

# 정확도
accuracy = accuracy_score(y_test, y_pred_03)
print(f'SVM의 정확도는 {accuracy}') 

''' 이건 한 번 그래프로 그려볼 수 있을까 '''

SVM의 정확도는 0.9861111111111112


' 이건 한 번 그래프로 그려볼 수 있을까 '

**[학습04.] SGD Classifier**

In [89]:
''' 학습 셋팅 '''
from sklearn.linear_model import SGDClassifier
sgd_model = SGDClassifier()

# 학습 
sgd_model.fit(X_train, y_train)

# 예측
y_pred_04 = sgd_model.predict(X_test)

# 정확도
accuracy = accuracy_score(y_test, y_pred_04)
print(f'SGD Classifier의 정확도는 {accuracy}') 

SGD Classifier의 정확도는 0.9527777777777777


**[학습05.] Logistic Regression**

In [94]:
''' 학습 셋팅 '''
from sklearn.linear_model import LogisticRegression
logistic_model = LogisticRegression(max_iter=2000)   # max_iter값 증가

# 학습
logistic_model.fit(X_train, y_train)

# 예측
y_pred_05 = logistic_model.predict(X_test)

# 정확도
accuracy = accuracy_score(y_test, y_pred_05)
print(f'Logistic Regression의 정확도는 {accuracy}') 

Logistic Regression의 정확도는 0.9722222222222222


            
<br>

**모델 평가**

정확도(accuracy)를 비교할 경우,
**Support Vector Machine(SVM)**이 정확도 **98.6%**로 가장 적합한 모델로 보인다.  


<br>

정확도를 보완하는 지표로 **F1 score** 값을 추가로 살펴보자.

0부터 9까지의 손글씨를 분류하는 행위는 해당 숫자를 Negative로 판별하여 놓치는 경우(FN)와 다른 숫자를 해당 숫자로 오인하여 Positive로 판별하는 경우(FP) 둘 다 치명적인 오류가 될 수 있다고 본다. 

- *예를 들면, 손글씨로 쓴 전화번호를 인식한다고 쳤을 때 FN, FP 경우 둘다 우리는 연락할 방법이 없어진다.*

따라서 Recall과 Precision의 조화평균인 F1 score를 선택한다.

다음은 Confusion matrix에 의해 각 모델의 성능지표 값을 나타낸 report이다.
**F1 score**의 경우에도 **Support Vector Machine(SVM)**이 **0.99**로 가장 높은 것을 알 수 있다.




In [105]:
from sklearn.metrics import confusion_matrix
from sklearn.metrics import classification_report

# Decision Tree report
confusion_matrix(y_test, y_pred_01)
print(classification_report(y_test, y_pred_01))

              precision    recall  f1-score   support

           0       0.97      0.88      0.92        33
           1       0.81      0.79      0.80        28
           2       0.87      0.82      0.84        33
           3       0.79      0.91      0.85        34
           4       0.81      0.85      0.83        46
           5       0.91      0.83      0.87        47
           6       0.94      0.94      0.94        35
           7       0.81      0.85      0.83        34
           8       0.85      0.77      0.81        30
           9       0.80      0.88      0.83        40

    accuracy                           0.85       360
   macro avg       0.86      0.85      0.85       360
weighted avg       0.86      0.85      0.85       360



In [106]:
# Random Forest report
confusion_matrix(y_test, y_pred_02)
print(classification_report(y_test, y_pred_02))

              precision    recall  f1-score   support

           0       1.00      1.00      1.00        33
           1       0.93      1.00      0.97        28
           2       1.00      1.00      1.00        33
           3       1.00      1.00      1.00        34
           4       1.00      1.00      1.00        46
           5       0.96      0.98      0.97        47
           6       0.97      0.97      0.97        35
           7       0.97      0.97      0.97        34
           8       1.00      0.93      0.97        30
           9       0.97      0.95      0.96        40

    accuracy                           0.98       360
   macro avg       0.98      0.98      0.98       360
weighted avg       0.98      0.98      0.98       360



In [107]:
# SVM report
confusion_matrix(y_test, y_pred_03)
print(classification_report(y_test, y_pred_03))

              precision    recall  f1-score   support

           0       1.00      1.00      1.00        33
           1       1.00      1.00      1.00        28
           2       1.00      1.00      1.00        33
           3       1.00      1.00      1.00        34
           4       1.00      1.00      1.00        46
           5       0.98      0.98      0.98        47
           6       0.97      1.00      0.99        35
           7       0.97      0.97      0.97        34
           8       1.00      0.97      0.98        30
           9       0.95      0.95      0.95        40

    accuracy                           0.99       360
   macro avg       0.99      0.99      0.99       360
weighted avg       0.99      0.99      0.99       360



In [108]:
# SGD Classifier report
confusion_matrix(y_test, y_pred_04)
print(classification_report(y_test, y_pred_04))

              precision    recall  f1-score   support

           0       1.00      0.97      0.98        33
           1       0.93      1.00      0.97        28
           2       0.97      1.00      0.99        33
           3       0.89      1.00      0.94        34
           4       0.98      1.00      0.99        46
           5       0.96      0.91      0.93        47
           6       0.97      0.97      0.97        35
           7       1.00      0.97      0.99        34
           8       0.87      0.90      0.89        30
           9       0.94      0.82      0.88        40

    accuracy                           0.95       360
   macro avg       0.95      0.96      0.95       360
weighted avg       0.95      0.95      0.95       360



In [109]:
# Logistic Regression report
confusion_matrix(y_test, y_pred_05)
print(classification_report(y_test, y_pred_05))

              precision    recall  f1-score   support

           0       1.00      1.00      1.00        33
           1       0.97      1.00      0.98        28
           2       0.97      1.00      0.99        33
           3       0.97      0.97      0.97        34
           4       1.00      0.98      0.99        46
           5       0.92      0.94      0.93        47
           6       0.97      0.97      0.97        35
           7       1.00      0.97      0.99        34
           8       0.97      0.97      0.97        30
           9       0.97      0.95      0.96        40

    accuracy                           0.97       360
   macro avg       0.97      0.97      0.97       360
weighted avg       0.97      0.97      0.97       360



<br>


**[소결]**



5가지 모델을 통해 손글씨 이미지를 판별한 결과는 다음과 같다.

*   항목 추가
*   항목 추가
*   항목 추가
*   항목 추가



(여기 테이블은 넣을 수 없나???)


<br><br><br><br><br><br>



In [ ]:
""" [Frame Work] """

''' 데이터 셋팅 '''
# 필요한 모듈 import

# 데이터 준비

# 데이터 이해
  # Feature Data 지정하기
  # Label Data 지정하기
  # Target Names 출력해 보기
  # 데이터 Describe 해 보기

# 데이터 구분



''' 학습 '''
# Decision Tree

# Random Forest

# SVM

# SGD Classifier

# Logistic Regression



''' 평가 '''
# 평가
